paddle/fluid/memory/allocation/allocator.h

```diff
@@ -162,53 +134,48 @@ class Allocation {
   friend class Allocator;
 };
 
+using Allocation = pten::candidate::Allocation;
+using AllocationPtr = pten::candidate::Allocator::AllocationPtr;
+using DecoratedAllocationPtr =
+    std::unique_ptr<DecoratedAllocation,
+                    pten::candidate::Allocator::DeleterType>;
+
 // Base interface class of memory Allocator.
-class Allocator {
+class Allocator : public pten::candidate::Allocator {
  public:
-  virtual ~Allocator() {}
-
-  class AllocationDeleter {
-   public:
-    inline void operator()(Allocation* allocation) const {
-      Allocator* allocator = allocation->TopDecoratedAllocator();
-      allocator->Free(allocation);
-    }
-  };
-
-  using AllocationPtr = std::unique_ptr<Allocation, AllocationDeleter>;
+  static void DecoratedDelete(pten::candidate::Allocation* allocation) {
+    auto decorated_allocation = static_cast<DecoratedAllocation*>(allocation);
+    Allocator* allocator = decorated_allocation->TopDecoratedAllocator();
+    allocator->DecoratedFree(decorated_allocation);
+  }
 
   // Allocate an allocation.
   // size may be 0, but it would be too complex if we handle size == 0
   // in each Allocator. So we handle size == 0 inside AllocatorFacade
   // in our design.
-  inline AllocationPtr Allocate(size_t size) {
+  AllocationPtr Allocate(size_t size) override {
+    return AllocationPtr(DecoratedAllocate(size), DecoratedDelete);
+  }
+
+  DecoratedAllocation* DecoratedAllocate(size_t size) {
```

Comment:
Is it good to return the `DecoratedAllocation*` in public method?

Reply:
`Allocator::DecoratedAllocate()` is used for situations where the allocator needs to acquire memory resources in the form of raw pointers. The original impl is that `Allocator::AllocImpl()` calls `std::unique_ptr::release()` and gets `Allocation*`. To completely solve the problem of public functions, it may be necessary to avoid `Allocator::Alloc` calls inside Allocator.